runtime/lib/Handlers/api/utils.ts

```diff
@@ -43,90 +59,160 @@ export const getVariable = (path: string, data: any) => {
   return stringToNumIfNumeric(curData);
 };
 
-export const ReduceKeyValue = (values: { key: string; val: string }[]) =>
-  values.reduce<Record<string, string>>((acc, { key, val }) => {
-    if (key) {
-      acc[key] = val;
-    }
-    return acc;
-  }, {});
+const validateUrl = async (urlString: string, runtime: Runtime) => {
+  let url;
+  try {
+    url = new URL(urlString);
+  } catch (err) {
+    throw new VError(`url: ${urlString} could not be parsed`, VError.HTTP_STATUS.BAD_REQUEST);
+  }
+
+  const { hostname } = url;
+
+  if (hostname.toLowerCase() === 'localhost') {
+    throw new VError(`url hostname cannot be localhost: ${urlString}`, VError.HTTP_STATUS.BAD_REQUEST);
+  }
+
+  // DNS.resolve returns an array of ips
+  const ips = validator.isIP(hostname)
+    ? [hostname]
+    : await DNS.resolve(hostname).catch(() => {
+        throw new VError(`cannot resolve hostname: ${hostname}`, VError.HTTP_STATUS.BAD_REQUEST);
+      });
+
+  PROHIBITED_IP_RANGES.forEach((prohibitedRange) => {
+    ips.forEach((ip) => {
+      if (ipRangeCheck(ip, prohibitedRange)) {
+        throw new VError(`url resolves to IP: ${ip} in prohibited range`, VError.HTTP_STATUS.BAD_REQUEST);
+      }
+    });
+  });
```

Comment:
I think using some more specific array methods can help clarify this piece of code:

```suggestion
  const badIP = ips.find((ip) => PROHIBITED_IP_RANGES.some((range) => ipRangeCheck(ip, range)));
  if (badIP) {
    throw new VError(`url resolves to IP: ${badIP} in prohibited range`, VError.HTTP_STATUS.BAD_REQUEST);
  }
```